File: node/astlib.cpp
#include <v8.h>
#include <node.h>
#include <node_buffer.h>

#include <v8pp/convert.hpp>
#include <v8pp/module.hpp>
#include <v8pp/class.hpp>
#include <v8pp/function.hpp>
#include <v8pp/object.hpp>

#include "astlib/SimpleAsterixRecord.h"
#include "astlib/CodecRegister.h"
#include "astlib/Exception.h"
#include "astlib/decoder/SimpleValueDecoder.h"
#include "astlib/decoder/BinaryAsterixDecoder.h"
#include "astlib/encoder/SimpleValueEncoder.h"
#include "astlib/encoder/BinaryAsterixEncoder.h"

#include <type_traits> // C++0x
#include <exception>
#include <iostream>

static int itemType(int value)
{
    return (value>>24) & 0xF;
}

template<typename T>
T fromV8Value(v8::Handle<v8::Value> value)
{
    if (std::is_same<T, std::string>::value)
    {
        v8::String::Utf8Value stringValue(value);
        return std::string(*stringValue, stringValue.length());
    }
    else if (std::is_same<T, int>::value)
        return value->IntegerValue();
    else
        throw new std::bad_cast();
}

std::string fromV8String(v8::Handle<v8::Value> value)
{
    v8::String::Utf8Value stringValue(value);
    return std::string(*stringValue, stringValue.length());
}

/*
v8::Handle<v8::Value> toV8Value(v8::Isolate* isolate, std::string& value)
{
    return v8::String::New(isolate, value.c_str(), value.length());
}

v8::Handle<v8::Value> toV8Object(v8::Isolate* isolate, int value)
{
    return v8::Number::New(isolate, value);
}
*/
template<typename T>
class Wrapper
{
public:
    explicit Wrapper(const v8::FunctionCallbackInfo<v8::Value>& args)
    {
    }

    T value;
};

using AsterixRecordWrapper = Wrapper<astlib::SimpleAsterixRecord>;
using CodecWrapper = Wrapper<astlib::CodecDescriptionPtr>;

static astlib::CodecRegister codecRegister;
static astlib::CodecDescriptionVector codecDescriptions;

static void loadCodecs()
{
    if (codecDescriptions.empty())
    {
        codecRegister.populateCodecsFromDirectory("../specs");
        codecDescriptions = codecRegister.enumerateAllCodecsByCategory();
    }
}

// TODO init kde sa nacitaju kodeky, error ak ziadny nenajde

v8::Handle<v8::Object> createAsterixRecord(const v8::FunctionCallbackInfo<v8::Value>& args)
{
    AsterixRecordWrapper* obj = new AsterixRecordWrapper(args);
    return v8pp::class_<AsterixRecordWrapper>::import_external(args.GetIsolate(), obj);
}

v8::Handle<v8::Array> enumerateAllCodecs()
{
    loadCodecs();

    v8::Isolate* isolate = v8::Isolate::GetCurrent();
    // We will be creating temporary handles so we use a handle scope.
    v8::EscapableHandleScope handleScope(isolate);
    v8::Local < v8::Array > array = v8::Array::New(isolate, 2);

    int index = 0;
    for (auto codec : codecDescriptions)
    {
        array->Set(index++, v8::String::NewFromUtf8(isolate, codec->getCategoryDescription().toString().c_str()));
    }

    return handleScope.Escape(array);
}

// AsterixRecord[] decodeAsterixBuffer(codecName, Buffer, Policy);
v8::Handle<v8::Value> decodeAsterixBuffer(const v8::FunctionCallbackInfo<v8::Value>& args)
{
    loadCodecs();

    v8::Isolate* isolate = v8::Isolate::GetCurrent();
    v8::EscapableHandleScope handleScope(isolate);

    std::string fullName = fromV8String(args[0]);

    astlib::CodecDescriptionPtr codec = codecRegister.getCodecForSignature(fullName);

    if (codec)
    {
        class MySimpleValueDecoder: public astlib::SimpleValueDecoder
        {
        public:
            virtual void onMessageDecoded(astlib::SimpleAsterixRecordPtr ptr)
            {
                msgs.push_back(ptr);
            }

            std::vector<astlib::SimpleAsterixRecordPtr> msgs;
        } myDecoder;

        astlib::Byte* buf = (astlib::Byte*) node::Buffer::Data(args[1]);
        size_t bytes = node::Buffer::Length(args[1]);
std::cout << "dec " << fullName << " len " << bytes << std::endl;
        astlib::BinaryAsterixDecoder decoder;

        try
        {
            decoder.decode(*codec, myDecoder, buf, bytes);
        }
        catch(astlib::Exception& e)
        {
            std::cerr << e.displayText() << std::endl;
        }

        if (myDecoder.msgs.size())
        {
            //zakoduj pole
        }
    }

    return handleScope.Escape(v8::Null(isolate)); // v8::Object::New(isolate)
}

// Buffer encodeAsterixRecord(codecName, AsterixRecord, Policy);
v8::Handle<v8::Value> encodeAsterixRecord(const v8::FunctionCallbackInfo<v8::Value>& args)
{
    loadCodecs();

    v8::Isolate* isolate = v8::Isolate::GetCurrent();
    v8::EscapableHandleScope handleScope(isolate);

    std::string fullName = fromV8String(args[0]);

    astlib::CodecDescriptionPtr codec = codecRegister.getCodecForSignature(fullName);

    if (codec)
    {
        std::vector<astlib::Byte> buffer;
        astlib::BinaryAsterixEncoder encoder;
/*
        try
        {
            astlib::SimpleValueEncoder valueEncoder;
            encoder.encode(*codec, valueEncoder, buffer);
        }
        catch(astlib::Exception& e)
        {
            std::cerr << e.displayText() << std::endl;
        }
        */
    }

    return handleScope.Escape(v8::Null(isolate)); // v8::Object::New(isolate)
}

std::string toString(AsterixRecordWrapper& obj)
{
    std::string value(obj.value.toString());
    return value;
}

bool hasItem(AsterixRecordWrapper& obj, int code)
{
    return obj.value.hasItem(code);
}

void allocateArray(AsterixRecordWrapper& obj, int code, int size)
{
    return obj.value.initializeArray(code, size);
}

double getNumberAt(AsterixRecordWrapper& obj, int code, int index)
{
    double value;
    Poco::Int64 value1 = 0;
    Poco::UInt64 value2 = 0;

    switch(itemType(code))
    {
        case astlib::PrimitiveType::Real:
            obj.value.getReal(code, value, index);
            break;

        case astlib::PrimitiveType::Integer:
            obj.value.getSigned(code, value1, index);
            value = value1;
            break;

        case astlib::PrimitiveType::Unsigned:
            obj.value.getUnsigned(code, value2, index);
            value = value2;
            break;

        default:
            throw std::bad_cast();
    }

    return value;
}

double getNumber(AsterixRecordWrapper& obj, int code)
{
    return getNumberAt(obj, code, -1);
}

void setNumberAt(AsterixRecordWrapper& obj, int code, double value, int index)
{
    switch(itemType(code))
    {
        case astlib::PrimitiveType::Real:
            obj.value.setItem(code, value, index);
            break;

        case astlib::PrimitiveType::Integer:
            obj.value.setItem(code, Poco::Int64(value), index);
            break;

        case astlib::PrimitiveType::Unsigned:
            obj.value.setItem(code, Poco::UInt64(value), index);
            break;

        default:
            throw std::bad_cast();
    }
}

void setNumber(AsterixRecordWrapper& obj, int code, double value)
{
    setNumberAt(obj, code, value, -1);
}

void setString(AsterixRecordWrapper& obj, int code, std::string value)
{
    obj.value.setItem(code, value);
}

void setStringAt(AsterixRecordWrapper& obj, int code, std::string value, int index)
{
    obj.value.setItem(code, value, index);
}

std::string getString(AsterixRecordWrapper& obj, int code)
{
    std::string value;
    obj.value.getString(code, value);
    return value;
}

std::string getStringAt(AsterixRecordWrapper& obj, int code, int index)
{
    std::string value;
    obj.value.getString(code, value, index);
    return value;
}

void setBoolean(AsterixRecordWrapper& obj, int code, bool value)
{
    obj.value.setItem(code, value);
}

bool getBoolean(AsterixRecordWrapper& obj, int code)
{
    bool value = false;
    obj.value.getBoolean(code, value);
    return value;
}

void setBooleanAt(AsterixRecordWrapper& obj, int code, bool value, int index)
{
    obj.value.setItem(code, value, index);
}

bool getBooleanAt(AsterixRecordWrapper& obj, int code, int index)
{
    bool value = false;
    obj.value.getBoolean(code, value, index);
    return value;
}

// TODO: ostatne gettery pre record

void InitAll(v8::Handle<v8::Object> exports)
{
    v8::Isolate* isolate = v8::Isolate::GetCurrent();

    v8pp::class_<AsterixRecordWrapper> MyObject_class(isolate);

    v8pp::module addon(isolate);
    addon.set("MyObject", MyObject_class);
    addon.set("createAsterixRecord", &createAsterixRecord);
    addon.set("enumerateAllCodecs", &enumerateAllCodecs);

    addon.set("decodeAsterixBuffer", &decodeAsterixBuffer);
    addon.set("encodeAsterixRecord", &encodeAsterixRecord);

    addon.set("hasItem", &hasItem);
    addon.set("allocateArray", &allocateArray);
    addon.set("setNumber", &setNumber);
    addon.set("getNumber", &getNumber);
    addon.set("setNumberAt", &setNumberAt);
    addon.set("getNumberAt", &getNumberAt);
    addon.set("setString", &setString);
    addon.set("getString", &getString);
    addon.set("setStringAt", &setStringAt);
    addon.set("getStringAt", &getStringAt);
    addon.set("setBoolean", &setBoolean);
    addon.set("getBoolean", &getBoolean);
    addon.set("setBooleanAt", &setBooleanAt);
    addon.set("getBooleanAt", &getBooleanAt);
    addon.set("toString", &toString);

    exports->SetPrototype(addon.new_instance());

    node::AtExit([](void* param)
    {
        v8pp::cleanup(static_cast<v8::Isolate*>(param));
    }, isolate);
}

NODE_MODULE(addon, InitAll)

File: node/astlib.js
//
// Copyright (c) 2013-2016 Pavel Medvedev. All rights reserved.
//
// This file is part of v8pp (https://github.com/pmed/v8pp) project.
//
// Distributed under the Boost Software License, Version 1.0. (See accompanying
// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
//

var assert = require("assert");
var astlib = require('./build/Release/addon.node');
const ASTERIX = require('./asterixitems.js');

var asterixRecord = astlib.createAsterixRecord();
var codecs = astlib.enumerateAllCodecs();

describe('AsterixRecord', function() {
	describe('#create', function() {
		it('"assert exists', function() {
			assert.ok(asterixRecord != null);			
		});
	}); 	

	describe('#set/get', function() {
		it('Real Number', function() {
			var pi = 3.14;
			assert(astlib.hasItem(asterixRecord, ASTERIX.TRACK_POSITION_LATITUDE) == false);
			astlib.setNumber(asterixRecord, ASTERIX.TRACK_POSITION_LATITUDE, pi);
			assert(astlib.hasItem(asterixRecord, ASTERIX.TRACK_POSITION_LATITUDE) == true);	
			assert(astlib.getNumber(asterixRecord, ASTERIX.TRACK_POSITION_LATITUDE) == pi);
		});

		it('Real Number Array', function() {
			var pi2 = [ 3.14, 6.28 ];
			assert(astlib.hasItem(asterixRecord, ASTERIX.AICRAFT_DERIVED_LAT) == false);
			astlib.allocateArray(asterixRecord, ASTERIX.AICRAFT_DERIVED_LAT, 2);
			assert(astlib.hasItem(asterixRecord, ASTERIX.AICRAFT_DERIVED_LAT) == true);
			astlib.setNumberAt(asterixRecord, ASTERIX.AICRAFT_DERIVED_LAT, pi2[0], 0);
			assert(astlib.getNumberAt(asterixRecord, ASTERIX.AICRAFT_DERIVED_LAT, 0) == pi2[0]);
			
			astlib.setNumberAt(asterixRecord, ASTERIX.AICRAFT_DERIVED_LAT, pi2[1], 1);
			assert(astlib.hasItem(asterixRecord, ASTERIX.AICRAFT_DERIVED_LAT) == true);	
			assert(astlib.getNumberAt(asterixRecord, ASTERIX.AICRAFT_DERIVED_LAT, 1) == pi2[1]);
		});

		it('Integer Number', function() {
			var i = -36745893;
			assert(astlib.hasItem(asterixRecord, ASTERIX.MODE5_GA) == false);
			astlib.setNumber(asterixRecord, ASTERIX.MODE5_GA, i);
			assert(astlib.hasItem(asterixRecord, ASTERIX.MODE5_GA) == true);	
			assert(astlib.getNumber(asterixRecord, ASTERIX.MODE5_GA) == i);
		});

		it('Integer Number Array', function() {
			var i2 = [ 3, -6 ];
			assert(astlib.hasItem(asterixRecord, ASTERIX.AICRAFT_DERIVED_TCPNUM) == false);
			astlib.allocateArray(asterixRecord, ASTERIX.AICRAFT_DERIVED_TCPNUM, 2);
			assert(astlib.hasItem(asterixRecord, ASTERIX.AICRAFT_DERIVED_TCPNUM) == true);
			astlib.setNumberAt(asterixRecord, ASTERIX.AICRAFT_DERIVED_TCPNUM, i2[0], 0);
			assert(astlib.getNumberAt(asterixRecord, ASTERIX.AICRAFT_DERIVED_TCPNUM, 0) == i2[0]);
			
			astlib.setNumberAt(asterixRecord, ASTERIX.AICRAFT_DERIVED_TCPNUM, i2[1], 1);
			assert(astlib.getNumberAt(asterixRecord, ASTERIX.AICRAFT_DERIVED_TCPNUM, 1) == i2[1]);
		});

		it('Unsigned Number', function() {
			var i = 54346745893;
			assert(astlib.hasItem(asterixRecord, ASTERIX.TRACK_STATUS_EMERGENCY) == false);
			astlib.setNumber(asterixRecord, ASTERIX.TRACK_STATUS_EMERGENCY, i);
			assert(astlib.hasItem(asterixRecord, ASTERIX.TRACK_STATUS_EMERGENCY) == true);	
			assert(astlib.getNumber(asterixRecord, ASTERIX.TRACK_STATUS_EMERGENCY) == i);
		});

		it('Unsigned Number Array', function() {
			var i2 = [ 34324563466, 67457456375 ];
			assert(astlib.hasItem(asterixRecord, ASTERIX.AICRAFT_DERIVED_TD) == false);
			astlib.allocateArray(asterixRecord, ASTERIX.AICRAFT_DERIVED_TD, 2);
			assert(astlib.hasItem(asterixRecord, ASTERIX.AICRAFT_DERIVED_TD) == true);
			astlib.setNumberAt(asterixRecord, ASTERIX.AICRAFT_DERIVED_TD, i2[0], 0);
			assert(astlib.getNumberAt(asterixRecord, ASTERIX.AICRAFT_DERIVED_TD, 0) == i2[0]);
			
			astlib.setNumberAt(asterixRecord, ASTERIX.AICRAFT_DERIVED_TD, i2[1], 1);
			assert(astlib.getNumberAt(asterixRecord, ASTERIX.AICRAFT_DERIVED_TD, 1) == i2[1]);
		});

		it('String', function() {
			var str = "Janťžščťžt";
			assert(astlib.hasItem(asterixRecord, ASTERIX.FPL_DATA_STAR) == false);	
			astlib.setString(asterixRecord, ASTERIX.FPL_DATA_STAR, str);
			assert(astlib.hasItem(asterixRecord, ASTERIX.FPL_DATA_STAR) == true);	
			assert(astlib.getString(asterixRecord, ASTERIX.FPL_DATA_STAR) == str);
		});

		it('String Number Array', function() {
			var s2 = [ "54363", "67457456375" ];
			assert(astlib.hasItem(asterixRecord, ASTERIX.MODES_MBDATA) == false);
			astlib.allocateArray(asterixRecord, ASTERIX.MODES_MBDATA, 2);
			assert(astlib.hasItem(asterixRecord, ASTERIX.MODES_MBDATA) == true);
			astlib.setStringAt(asterixRecord, ASTERIX.MODES_MBDATA, s2[0], 0);
			assert(astlib.getStringAt(asterixRecord, ASTERIX.MODES_MBDATA, 0) == s2[0]);
			
			astlib.setStringAt(asterixRecord, ASTERIX.MODES_MBDATA, s2[1], 1);
			assert(astlib.getStringAt(asterixRecord, ASTERIX.MODES_MBDATA, 1) == s2[1]);
		});

		it('Boolean', function() {
			var onoff = true;
			assert(astlib.hasItem(asterixRecord, ASTERIX.ACAS_RESOLUTION_ARA42) == false);	
			astlib.setBoolean(asterixRecord, ASTERIX.ACAS_RESOLUTION_ARA42, onoff);
			assert(astlib.hasItem(asterixRecord, ASTERIX.ACAS_RESOLUTION_ARA42) == true);	
			assert(astlib.getBoolean(asterixRecord, ASTERIX.ACAS_RESOLUTION_ARA42) == onoff);
		});

		/*
		 *  NO DEFINED BOOLEAN ARRAY YET
		it('Boolean Number Array', function() {
			var s2 = [ true, false ];
			assert(astlib.hasItem(asterixRecord, ASTERIX.MODES_MBDATA) == false);
			astlib.allocateArray(asterixRecord, ASTERIX.MODES_MBDATA, 2);
			assert(astlib.hasItem(asterixRecord, ASTERIX.MODES_MBDATA) == true);
			astlib.setBooleanAt(asterixRecord, ASTERIX.MODES_MBDATA, s2[0], 0);
			assert(astlib.getBooleanAt(asterixRecord, ASTERIX.MODES_MBDATA, 0) == s2[0]);
			
			astlib.setBooleanAt(asterixRecord, ASTERIX.MODES_MBDATA, s2[1], 1);
			assert(astlib.getBooleanAt(asterixRecord, ASTERIX.MODES_MBDATA, 1) == s2[1]);
		});
		 */
		it('Stringify', function() {
			assert.equal(astlib.toString(asterixRecord).length, 309);	
		});
	
		// TODO: JSONify
	}); 	
});


describe('Asterix Encoder', function() {	
	var codecs = astlib.enumerateAllCodecs();
	
	describe('#encode', function() {
		it('empty record', function() {
			//console.log(astlib.toString(asterixRecord));
			var buffer = astlib.encodeAsterixRecord(asterixRecord, codecs[0]);			
			assert.ok(buffer != null);			
			assert.equal(buffer.length, 3);
		});
		
		it('cat. 48 plot', function() {
			var plot = astlib.createAsterixRecord();
			astlib.setNumber(plot, ASTERIX.DSI_SAC, 44);
			astlib.setNumber(plot, ASTERIX.DSI_SIC, 144);
			astlib.setNumber(plot, ASTERIX.TIMEOFDAY, 3600);
			astlib.setNumber(plot, ASTERIX.TRACK_DETECTION, 3);
			astlib.setBoolean(plot, ASTERIX.TRACK_TEST, true);
			astlib.setBoolean(plot, ASTERIX.TRACK_SIMULATED, true);
			
			astlib.setNumber(plot, ASTERIX.TRACK_POSITION_RANGE, 10000.0);
			astlib.setNumber(plot, ASTERIX.TRACK_POSITION_AZIMUTH, 45.0);
			
			astlib.setBoolean(plot, ASTERIX.MODE3A_V, true);
			astlib.setBoolean(plot, ASTERIX.MODE3A_G, true);
			astlib.setBoolean(plot, ASTERIX.MODE3A_L, true);
			astlib.setNumber(plot, ASTERIX.MODE3A_VALUE, 9053);
			console.log(astlib.toString(plot));
			var buffer = astlib.encodeAsterixRecord(plot, 'Eurocontrol-48:1.21');			
			assert.ok(buffer != null);			
			assert.equal(buffer.length, 17);
			console.log(buffer);
			
			var record = astlib.decodeAsterixBuffer('Eurocontrol-48:1.21', buffer);
			console.log(astlib.toString(record[0]));
		});
	}); 	
});

